File: 0561-array-partition/0561-array-partition.cpp
class Solution {
public:
    int arrayPairSum(vector<int>& nums) {
        int maxVal = nums[0], minVal = nums[0];
        int n = nums.size();

        for (int i = 1; i < n; i++) {
            if (maxVal < nums[i]) maxVal = nums[i];
            if (minVal > nums[i]) minVal = nums[i];
        }

        int l = maxVal - minVal + 1;
        vector<int> c(l, 0);

        for (int i = 0; i < n; i++) {
            c[nums[i] - minVal]++;
        }

        int sum = 0;
        bool s = false;

        for (int k = 0; k < l; k++) {
            if (c[k] != 0) {
                int a = c[k];

                if (s) a--;

                sum += ((a + 1) / 2) * (k + minVal);

                s = a % 2 == 1;
            }
        }

        return sum;
    }
};

File: 0561-array-partition/0561-array-partition.rs
impl Solution {
    pub fn array_pair_sum(nums: Vec<i32>) -> i32 {
        let mut max_val = nums[0];
        let mut min_val = nums[0];

        for i in 1..nums.len() {
            if max_val < nums[i] {
                max_val = nums[i];
            }

            if min_val > nums[i] {
                min_val = nums[i];
            }
        }

        let l = (max_val - min_val + 1) as usize;
        let mut c = vec![0; l];

        for &x in &nums {
            c[(x - min_val) as usize] += 1;
        }

        let mut sum = 0;
        let mut s = false;

        for k in 0..l {
            if c[k] != 0 {
                let mut a = c[k];

                if s {
                    a -= 1;
                }

                sum += ((a + 1) / 2) * (k as i32 + min_val);

                s = a % 2 == 1;
            }
        }

        sum
    }
}